syntheseus/reaction_prediction/utils/inference.py

```diff
@@ -1,32 +1,73 @@
 from pathlib import Path
-from typing import Any, Dict, List, Sequence, Union
-
-from syntheseus.interface.models import Prediction
+from typing import Any, Dict, List, Sequence, Union, cast
+
+from syntheseus.interface.bag import Bag
+from syntheseus.interface.molecule import Molecule
+from syntheseus.interface.reaction import (
+    MultiProductReaction,
+    ReactionMetaData,
+    SingleProductReaction,
+)
 from syntheseus.reaction_prediction.chem.utils import molecule_bag_from_smiles
 
 
-def process_raw_smiles_outputs(
-    input: Any, output_list: List[str], kwargs_list: List[Dict[str, Any]]
-) -> Sequence[Prediction]:
-    """Convert raw SMILES outputs into a list of `Prediction` objects.
+def process_raw_smiles_outputs_backwards(
+    input: Molecule, output_list: List[str], metadata_list: List[Dict[str, Any]]
+) -> Sequence[SingleProductReaction]:
+    """Convert raw SMILES outputs into a list of `SingleProductReaction` objects.
 
     Args:
-        inputs: Model input (can be `Molecule` or `Bag[Molecule]` depending on directionality).
+        input: Model input.
         output_list: Raw SMILES outputs (including potentially invalid ones).
         kwargs_list: Additional metadata to attach to the predictions (e.g. probability).
 
     Returns:
-        A list of `Prediction`s; may be shorter than `outputs` if some of the raw
+        A list of `SingleProductReaction`s; may be shorter than `outputs` if some of the raw
         SMILES could not be parsed into valid reactant bags.
     """
-    predictions: List[Prediction] = []
+    predictions: List[SingleProductReaction] = []
 
-    for raw_output, kwargs in zip(output_list, kwargs_list):
+    for raw_output, metadata in zip(output_list, metadata_list):
         reactants = molecule_bag_from_smiles(raw_output)
 
         # Only consider the prediction if the SMILES can be parsed.
         if reactants is not None:
-            predictions.append(Prediction(input=input, output=reactants, **kwargs))
+            predictions.append(
+                SingleProductReaction(
+                    product=input, reactants=reactants, metadata=cast(ReactionMetaData, metadata)
```

Comment:
Does using a plain `dict` casted to `ReactionMetaData` instead of an actual instance of `ReactionMetaData` have any non-typing-related consequences? Or is there nothing a typed dict can do that a plain dict cannot outside of being typed better?

Reply:
A typed dict is just for type-hinting. The [docs](https://docs.python.org/3/library/typing.html#typing.TypedDict) say that "At runtime it is a plain dict". The cast is just because some model-specific keys are added which are not typed in ReactionMetadata.